src/test/ui/async-await/async-fn-size-moved-locals.rs

```diff
@@ -112,7 +112,7 @@ async fn mixed_sizes() {
 fn main() {
     assert_eq!(1025, std::mem::size_of_val(&single()));
     assert_eq!(1026, std::mem::size_of_val(&single_with_noop()));
-    assert_eq!(3078, std::mem::size_of_val(&joined()));
-    assert_eq!(3079, std::mem::size_of_val(&joined_with_noop()));
-    assert_eq!(7181, std::mem::size_of_val(&mixed_sizes()));
+    assert_eq!(3076, std::mem::size_of_val(&joined()));
+    assert_eq!(3076, std::mem::size_of_val(&joined_with_noop()));
+    assert_eq!(6157, std::mem::size_of_val(&mixed_sizes()));
```

Comment:
This is unexpected! Do you know why these sizes change?

It looks like an improvement but I'd like to understand what's going on.

Reply:
We weren't sure either! The previous PR also [shrunk the size of the generated futures](https://github.com/rust-lang/rust/pull/65244/files#diff-629b10b8de5cb576c915da39e5d2a44042abfbaa33e3692d7125436015c50e10R115-R117), without really commenting on it either. It's particularly surprising because neither the numbers before or after this changed are aligned the way we expected them to.

We reasoned that this change would be alright since it doesn't regress, there is precedence for accepting this in https://github.com/rust-lang/rust/pull/65244, and it doesn't lead to any other changes as far as we can tell. Even if we don't fully understand why the current results are what they are, or what is causing the change.

Reply:
I don't know if this is right, but one guess is it has something to do with the `match <future> { ... }` becoming `match InfoFuture::into_future(<future>) { ... }`. Maybe the `into_future` call moves the future into the scrutinee position of the match, rather than borrowing from the outer scope?

Judging by the comment on this test, it seems like the first two changes aren't a big deal, since they are just a few bytes. The last change is surprising because it improves by exactly 1024. This means we were able to eliminate one of the `BigFut`s from the closure. By my count we should only need 5, so I'm not sure why we have 6 copies now or why we had 7 copies before. I wouldn't be surprised to see 8 or 5, I guess, but 6 and 7 are both weird. Could we be smarter about laying out the closures? Maybe there's a move from one slot to another that's forced by the way we laid out the closures?

Reply:
Interesting theory, if we're turning a borrow into a move that can definitely improve things! See #62321 for instance – if I recall correctly, borrowing a local in a generator causes the analysis to assume the local is always live from that point.